x509-cert/src/builder.rs

```diff
@@ -0,0 +1,373 @@
+        let signer_pub = signer.public_key();
+
+        let signature_alg = signer.signature_algorithm().identifier;
+        let issuer = profile.get_issuer(&subject);
+
+        validity.not_before.rfc5280_adjust_utc_time()?;
+        validity.not_after.rfc5280_adjust_utc_time()?;
+
+        let (version, (issuer_unique_id, subject_unique_id)) = match version {
+            CertificateVersion::V1 => (Version::V1, (None, None)),
+            CertificateVersion::V2(uids) => (Version::V2, uids.get_unique_ids()),
+            CertificateVersion::V3(uids) => (Version::V3, uids.get_unique_ids()),
+        };
+
+        let mut tbs = TbsCertificate {
+            version,
+            serial_number,
+            signature: signature_alg,
+            issuer,
+            validity,
+            subject,
+            subject_public_key_info,
+            issuer_unique_id,
+            subject_unique_id,
+            extensions: None,
+        };
+
+        if tbs.version == Version::V3 {
+            let extensions =
+                profile.build_extensions(&tbs.subject_public_key_info, &signer_pub, &tbs)?;
+            if !extensions.is_empty() {
+                tbs.extensions = Some(extensions);
+            }
+        }
+
+        Ok(Self { tbs, signer })
+    }
+
+    /// Add an extension to this certificate
+    pub fn add_extension<E: AsExtension>(&mut self, extension: &E) -> Result<()> {
+        if self.tbs.version == Version::V3 {
+            let ext = extension.to_extension(&self.tbs)?;
+
+            if let Some(extensions) = self.tbs.extensions.as_mut() {
+                extensions.push(ext);
+            } else {
+                let extensions = vec![ext];
+                self.tbs.extensions = Some(extensions);
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Run the certificate through the signer and build the end certificate.
+    pub fn build(&mut self) -> Result<core::result::Result<Certificate, S::Err>> {
+        let signature = match self.signer.sign(&self.tbs) {
+            Ok(s) => s,
+            Err(e) => return Ok(Err(e)),
+        };
+        let signature = BitString::from_bytes(&signature)?;
+
+        let cert = Certificate {
+            tbs_certificate: self.tbs.clone(),
+            signature_algorithm: self.tbs.signature.clone(),
+            signature,
+        };
+
+        Ok(Ok(cert))
+    }
+}
+
+/// Signer to be used to for signing the certificates
+pub trait Signer {
+    /// Error to be returned by the Signer
+    type Err: fmt::Debug;
+
+    /// The signature expected from this signer
+    fn signature_algorithm(&self) -> CertificateSignatureAlgorithmOwned;
+
+    /// The SPKI encoded public key used by this signer
+    fn public_key(&self) -> SubjectPublicKeyInfoOwned;
+
+    /// The sign method should return the signature of the payload.
+    // TODO(baloo): do we need to zeroize that?
+    fn sign(&mut self, input: &TbsCertificate) -> core::result::Result<vec::Vec<u8>, Self::Err>;
+}
```

Comment:
It'd be really nice if we can glue this into the existing traits in e.g. `signature` and `spki`. Otherwise, how is it supposed to get implemented? Either `x509-cert` needs to implement it for every key type, or downstream crates need to add `x509-cert` as a dependency and implement it.

I think it can be decomposed into the following:

- `signature_algorithm()` could potentially be provided by the newly added `(Dyn)AssociatedAlgorithmIdentifier` when impl'd on a signature type like `rsa::pkcs1::Signature` or `p256::ecdsa::Signature`, although that would return an `AlgorithmIdentifierOwned`, not a `CertificateSignatureAlgorithmOwned`.
- `public_key` can be provided by `spki::EncodePublicKey`
- `sign` can be provided by `signature::Signer`